Adding variable to define Custom browser to use

shubham-cpp opened the pull request:
Add an option for user to specify using different browser(other than default one) for org/markdown previews. Its working for me right now, but will need further  testing for any edge cases(if they exist).

Also updated README to include how to set the variable.

shubham-cpp committed:
Added functionality to specify custom browser

Added custom variable `grip-url-browser` to specify another browser to
be used instead of default one.

Added function grip--browser which will load preview either by using
grip-url-browser or default browser

shubham-cpp committed:
Documentation for grip-url-browser

seagle0128 commented on grip-mode.el:
```diff
@@ -98,6 +103,16 @@ option."
+Use default browser if nil."
+  (if grip-url-browser
+      (progn
+       (setq-local browse-url-generic-program grip-url-browser
```
Doesn't `let` work here? I prefer `let` than `setq-local`.

Or use this
```elisp
(when grip-url-browser
  (setq-local browse-url-generic-program grip-url-browser
              browse-url-browser-function 'browse-url-generic))
(browse-url url)
```

shubham-cpp replied:
> let


For some reason let is causing issues with void function. **when** works as expected

seagle0128 replied:
OK.

seagle0128 commented on grip-mode.el:
```diff
@@ -56,6 +56,12 @@
+(defcustom grip-url-browser nil
+  "Browser to launch Markdown/Org previews.
+Use default browser if nil."
+  :type 'string
```
Should be: `:type '(choice (const :tag "None" nil) string)`.

shubham-cpp replied:
Ok that makes sense

shubham-cpp committed:
Changed from let to when and update type of grip-url-browser

All the changes have been updated as you said. I also made sure that
these changes are working as expected.

Let was giving me symbol's defination is void function something Like
that so by searching online I found setq-local.

seagle0128 commented on grip-mode.el:
```diff
@@ -56,6 +56,12 @@
+Use default browser if nil."
+  :type '(choice (const :tag "None" nil) string)
+  :group 'grip)
+
```
I checked `browse-url.el` today, and I think another option is necessary.

```elisp
(defcustom grip-url-args nil
  "A list of strings defining options for `grip-url-browser'."
  :type '(repeat (string :tag "Argument")))
```

seagle0128 commented on grip-mode.el:
```diff
@@ -98,6 +103,14 @@ option."
+Use default browser if nil."
+  (when grip-url-browser
+    (setq-local browse-url-generic-program grip-url-browser
+                browse-url-browser-function 'browse-url-generic))
```
Also set `grip-url-args`.

shubham-cpp replied:
will look into it

seagle0128 replied:
Just use

```elisp
(setq-local browse-url-generic-program grip-url-browser
            browse-url-generic-args grip-url-args
            browse-url-browser-function 'browse-url-generic)
```

shubham-cpp replied:
But what if user wants to pass arguments to default browser. I also though just add it to setq-local, but then it came to mind that maybe user wants to pass args to regular browser as well

seagle0128 replied:
So the best solution should be use `let`. The local vars won't be override. 
I tried this snippet and it works as expected.

```elisp
(defun grip--browser (url)
  "Use browser specified by user to load URL.
Use default browser if nil."
  (if grip-url-browser
      (let ((browse-url-generic-program grip-url-browser)
            (browse-url-generic-args grip-url-args))
        (ignore browse-url-generic-program)
        (ignore browse-url-generic-args)
        (browse-url-generic url))
    (browse-url url)))
```

shubham-cpp replied:
yes. tested it. works as expected

seagle0128 reviewed (changes requested):
Overall it looks good now. Please make the changes, and then I will approve and merge. Thanks!

shubham-cpp committed:
Added grip-url-args

shubham-cpp committed:
Added grip-url-args

shubham-cpp committed:
updated readme to include  documentation

shubham-cpp committed:
using let instead of setq-local

seagle0128 commented:
Thank you!
